### ISSUE TITLE
[FEATURE REQUEST] Always do squash of all commits in pull requests on merge action

### ISSUE THREAD

**Issue by MurzNN** [saltstack/salt#61653]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when see a lot of useless commits with messages like [pre commit](https://github.com/saltstack/salt/commit/f791398d9cedde0c8528a2cd8451d4eb2a60b9a4), [fix tests](https://github.com/saltstack/salt/commit/55a7519dd5dab2bdfcac2e7e6e77a3d1358538f9), [Add missing import](https://github.com/saltstack/salt/commit/21f7658ae42e85fbc51c2a5509cf2c340f40a317), that comes from various pull requests and stays as separate commits in main branch.

**Describe the solution you'd like**
I want to suggest to always do a squash of pull requests on merge action, to join all PR commits into one with message text from the PR title.

So, after PR with 25 commits will be merged, it'll add to `master` branch histor only one single combined commit with all changes from PR, and with nice clear message from PR title and optionally a description too.

This will make a list of commits in `master` branch looks much better, without bunch of "typo fix"-like commits, and easier to find reasons for concrete changes.

**Describe alternatives you've considered**
Alternative can be asking PR authors to manually squash all commits into one before merging, but I think that it's easier to automate on merge action, instead of poking each author to do this.

**Comment by baby-gnu** [comment]
Hello.

I think that instead of squashing commits, the PR authors should clean their branches before the final merge.

When I read the logs of a project, I start with `git log --oneline --merges`.

So, when I do a PR, I have 2 habits.

First, I create a meaningful branch name for the `git log --oneline --merges`

Then, I generally reduce my PR to several commits:

1. a dedicated commit to add/modify the documentation to explain the changes
2. a dedicated commit to add tests, when applicable
3. the commit to make the tests pass

So, the `git log --oneline --merges` permits to have a quick look of what's happening in a project, then if I'm interested by what does a contribution, I read the logs corresponding to that commit.

Regards.

**Comment by OrangeDog** [comment]
There is often benefit to multiple commits in a PR, usually as the complexity grows.

What gets merged should be tidy, but not forced to always be a single commit.
It's up to PR authors to do it, and the people merging to enforce it.

**Comment by MurzNN** [comment]
Yes, sometimes commits squashing is ok, sometimes - not, so let's describe this as recommendation? Author of PR can choose if he wants to squash commits on merge, or not.

For me it's very good to see a single commit in `main` branch from PR, with keeping ability to see all separate commits in PR branch, if needed. So asking authors to squash their commits manually for PR will remove this cool ability :(

**Comment by baby-gnu** [comment]
> For me it's very good to see a single commit in `main` branch from PR, with keeping ability to see all separate commits in PR branch, if needed for some detailed analyze. So asking authors to squash their commits manually in their branches for PR will remove this cool ability :(

Unfortunately, the PR branch is often removed when merged, so squashing in a single commit loose information.

Viewing a single commit in `main` branch is done by `git log --oneline --merges`. It's a shame that the web frontend of github does not permit that kind of display but I mostly use [the plugin of my editor](https://magit.vc/) so I don't worry about github limitations.
